src/dmd/cppmangle.d

```diff
@@ -1271,6 +1359,112 @@ private final class CppMangleVisitor : Visitor
     }
 
+    /**
+     * Helper function to write a `T..._` template index.
+     *
+     * Params:
+     *   ident = Identifier for which substitution is attempted
+     *           (e.g. `void func(T)(T param)` => `T` from `T param`)
+     *   params = `TemplateParameters` of the enclosing symbol
+     *           (in the previous example, `func`'s template parameters)
+     *   type = Resolved type of `T`, so that `void func(T)(const T)`
+     *          gets mangled correctly
+     *
+     * Returns:
+     *   `true` if something was written to the buffer
+     */
+    private void writeTemplateArgIndex(size_t idx, TemplateParameter param)
+    {
+        // expressions are mangled in <X..E>
+        if (param.isTemplateValueParameter())
+            buf.writeByte('X');
+        buf.writeByte('T');
+        writeSequenceFromIndex(idx);
+        buf.writeByte('_');
+        if (param.isTemplateValueParameter())
+            buf.writeByte('E');
+    }
+
+    /**
+     * Given an array of template parameters and an identifier,
+     * returns the index of the identifier in that array.
+     *
+     *
+     * Returns:
+     *   The index of the identifier match in `params`,
+     *   or `params.dim` if there wasn't any match.
+     */
+    private static size_t templateParamIndex(
+        const ref Identifier ident, TemplateParameters* params)
+    {
+        foreach (idx, param; *params)
+            if (param.ident == ident)
+                return idx;
+        return params.length;
+    }
+
+    /**
+     * Given a template instance `t`, write its qualified name
+     * without the template parameter list
+     *
+     * Params:
+     *   t = Post-parsing `TemplateInstance` pointing to the symbol
+     *       to mangle (one level deep)
+     *   dg = Delegate to execute after writing the qualified symbol
+     *
+     */
+    private void writeQualified(TemplateInstance t, scope void delegate() dg)
+    {
+        auto type = isType(this.context.res);
+        if (!type)
+        {
+            this.writeIdentifier(t.name);
+            return dg();
+        }
+        auto sym1 = type.toDsymbol(null);
+        if (!sym1)
+        {
+            this.writeIdentifier(t.name);
+            return dg();
+        }
+        // Get the template instance
+        auto sym = getQualifier(sym1);
+        auto sym2 = getQualifier(sym);
+        if (sym2)
+        {
+            if (isStd(sym2))
+            {
+                bool unused;
+                assert(sym.isTemplateInstance());
+                if (this.writeStdSubstitution(sym.isTemplateInstance(), unused))
+                    return dg();
+                // std names don't require `N..E`
+                buf.writestring("St");
+                this.writeIdentifier(t.name);
+                this.append(t);
+                return dg();
+            }
+            buf.writestring("N");
+            if (!this.substitute(sym2))
+                sym2.accept(this);
+            dg();
```

Comment:
@Geod24 duplicated call. and missing terminator `"E"` for `"N"`.
It seems this branch isn't covered in the tests.
```d
extern(C++, A) struct S(T){}
extern(C++) void fun(int i)(S!int);
pragma(msg, fun!0.mangleof);
```
Prints `_Z3funILi0EEvN1AIiE1SIiE`. should be `_Z3funILi0EEvN1A1SIiEE`.